src/securemul.rs

```diff
@@ -0,0 +1,430 @@
+use crate::error::BoxError;
+use crate::field::Field;
+use crate::helpers::ring::{HelperAddr, Ring};
+use crate::prss::Participant;
+use crate::replicated_secret_sharing::ReplicatedSecretSharing;
+use serde::{Deserialize, Serialize};
+use std::fmt::Debug;
+use thiserror::Error;
+
+/// Secure multiplication protocol using replicated secret sharing over some field `F`.
+/// * [paper](https://eprint.iacr.org/2018/387.pdf)
+#[derive(Debug)]
+pub struct SecureMul<F> {
+    index: u128,
+    a_share: ReplicatedSecretSharing<F>,
+    b_share: ReplicatedSecretSharing<F>,
+}
+
+/// A message sent by each helper when they've multiplied their own shares
+#[derive(Debug, Serialize, Deserialize, Default)]
+pub struct DValue {
+    index: u128,
+    d: u128,
+}
+
+/// Context used by each helper to perform computation. Currently they need access to shared
+/// randomness generator (PRSS) and communication trait to send messages to each other.
+/// Eventually when we have more than one protocol, this should be lifted to its own module
+#[derive(Debug)]
+pub struct ProtocolContext<'a, R> {
+    pub participant: &'a Participant,
+    pub helper_ring: &'a R,
+}
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error(
+        "Shares calculated by peer used different index {their_index} than expected {my_index}"
+    )]
+    IndexMismatch { my_index: u128, their_index: u128 },
+}
+
+impl<F: Field> SecureMul<F> {
+    /// Executes the secure multiplication on the MPC helper side. Each helper will proceed with
+    /// their part, eventually producing 2/3 shares of the product and that is what this function
+    /// returns.
+    ///
+    /// ## Errors
+    /// Lots of things may go wrong here, from timeouts to bad output. They will be signalled
+    /// back via the error response
+    pub async fn execute<R: Ring>(
```

Comment:
I see that we lost the Optimization here. that's fine - it keeps this diff simpler. We can add it later.